repl/src/dotty/tools/repl/ParseResult.scala

```diff
@@ -0,0 +1,147 @@
+
+/** `:load <path>` interprets a scala file as if entered line-by-line into
+ *  the REPL
+ */
+case class Load(path: String) extends Command
+object Load {
+  val command = ":load"
+}
+
+/** To find out the type of an expression you may simply do:
+ *
+ * ```
+ * scala> :type (1 * 54).toString
+ * String
+ * ```
+ */
+case class TypeOf(expr: String) extends Command
+object TypeOf {
+  val command = ":type"
+}
+
+/** `:imports` lists the imports that have been explicitly imported during the
+ *  session
+ */
+case object Imports extends Command {
+  val command = ":imports"
+}
+
+/** Reset the session to the initial state from when the repl program was
+ *  started
+ */
+case object Reset extends Command {
+  val command = ":reset"
+}
+
+/** `:quit` exits the repl */
+case object Quit extends Command {
+  val command = ":quit"
+}
+
+/** `:help` shows the different commands implemented by the Dotty repl */
+case object Help extends Command {
+  val command = ":help"
+  val text =
+    """The REPL has several commands available:
+      |
+      |:help                    print this summary or command-specific help
+      |:load <path>             interpret lines in a file
+      |:quit                    exit the interpreter
+      |:type <expression>       evaluate the type of the given expression
+      |:reset                   reset the repl to its initial state, forgetting all session entries
+    """.stripMargin
+}
+
+object ParseResult {
+
+  private[this] val CommandExtract = """(:[\S]+)\s*(.*)""".r
+
+  /** Extract a `ParseResult` from the string `sourceCode` */
+  def apply(sourceCode: String)(implicit ctx: Context): ParseResult =
+    sourceCode match {
+      case "" => Newline
+      case CommandExtract(cmd, arg) => cmd match {
+        case Quit.command => Quit
+        case Help.command => Help
+        case Reset.command => Reset
+        case Imports.command => Imports
+        case Load.command => Load(arg)
+        case TypeOf.command => TypeOf(arg)
+        case _ => UnknownCommand(cmd)
+      }
+      case _ => {
+        val source = new SourceFile("<console>", sourceCode.toCharArray)
+        val parser = new Parser(source)
+
+        val (_, stats) = parser.templateStatSeq()
+
+        if (ctx.reporter.hasErrors) {
+          SyntaxErrors(sourceCode,
+                       ctx.flushBufferedMessages(),
+                       stats)
+        }
+        else
+          Parsed(sourceCode, stats)
+      }
+    }
+
+  /** Check if the input is incomplete
+   *
+   *  This can be used in order to check if a newline can be inserted without
+   *  having to evaluate the expression
+   */
+  def isIncomplete(sourceCode: String)(implicit ctx: Context): Boolean =
+    sourceCode match {
+      case CommandExtract(_) | "" => false
+      case _ => {
+        val reporter = storeReporter
+        var needsMore = false
+        reporter.withIncompleteHandler(_ => _ => needsMore = true) {
```

Comment:
As discussed 🤓 to 🤓 , this is so much simpler than changing the flow of `Parsers.scala`. So we're going to live with this 👿 mutation for now :+1: